lib/schemes/oauth2.js

```diff
@@ -171,4 +173,71 @@ export default class Oauth2Scheme {
 
     return true // True means a redirect happened
   }
+
+  // ---------------------------------------------------------------
+  // Watch axios requests for token expiration
+  // ---------------------------------------------------------------
+
+  watchTokenExpiration () {
+    const { $axios } = this.$auth.ctx.app
+    let isRefreshing = false
+
+    $axios.onRequest(async (config) => {
+      // token and refresh token
+      let token = this.$auth.getToken(this.name)
+      let refreshToken = this.$auth.getRefreshToken(this.name)
+
+      // token or "refresh token" does not exists
+      if (!token || !refreshToken || !token.length || !refreshToken.length) {
+        return config
+      }
+
+      // if already trying to refresh token, do not try again
+      if (isRefreshing) {
+        return config
+      }
+
+      // time variables
+      const tokenExpiresAt = jwtDecode(token).exp * 1000
+      const refreshTokenExpiresAt = jwtDecode(refreshToken).exp * 1000
+      const now = Date.now()
+
+      // token has not been expired
+      if (now < tokenExpiresAt) {
+        return config
+      }
+
+      // "refresh token" also has been expired. There is no way to refresh. Force logout.
+      if (now > refreshTokenExpiresAt) {
+        this.logout()
+      }
+
+      // Try to refresh token before processing current request
+      isRefreshing = true
+
+      return $axios.post(this.options.access_token_endpoint,
```

Comment:
Please use async-await and when the refresh fails, please log the user out.

Reply:
I think this is safe to return a promise.